firebase-app-distribution/src/main/java/com/google/firebase/app/distribution/FirebaseAppDistributionTesterApiClient.java

```diff
@@ -58,31 +56,69 @@
 
   public static final int DEFAULT_BUFFER_SIZE = 8192;
 
+  private final HttpsUrlConnectionFactory httpsUrlConnectionFactory;
+
+  FirebaseAppDistributionTesterApiClient() {
+    this(new HttpsUrlConnectionFactory());
+  }
+
+  @VisibleForTesting
+  FirebaseAppDistributionTesterApiClient(
+      @NonNull HttpsUrlConnectionFactory httpsUrlConnectionFactory) {
+    this.httpsUrlConnectionFactory = httpsUrlConnectionFactory;
+  }
+
   /**
    * Fetches and returns the lastest release for the app that the tester has access to, or null if
    * the tester doesn't have access to any releases.
    */
   @Nullable
-  public AppDistributionReleaseInternal fetchNewRelease(
+  AppDistributionReleaseInternal fetchNewRelease(
       @NonNull String fid,
       @NonNull String appId,
       @NonNull String apiKey,
       @NonNull String authToken,
       @NonNull Context context)
       throws FirebaseAppDistributionException {
-    HttpsURLConnection connection = openHttpsUrlConnection(appId, fid, apiKey, authToken, context);
+    HttpsURLConnection connection = null;
+    int responseCode;
     String responseBody;
-    try (BufferedInputStream inputStream = new BufferedInputStream(connection.getInputStream())) {
-      responseBody = convertInputStreamToString(inputStream);
+    try {
+      connection = openHttpsUrlConnection(appId, fid, apiKey, authToken, context);
+      responseCode = connection.getResponseCode();
+      responseBody = readResponseBody(connection);
     } catch (IOException e) {
-      throw getExceptionForHttpResponse(connection, e);
+      throw new FirebaseAppDistributionException(
+          ErrorMessages.NETWORK_ERROR, Status.NETWORK_FAILURE, e);
     } finally {
-      connection.disconnect();
+      if (connection != null) {
+        // TODO(lkellogg): check if we're disconnecting everywhere we should
+        connection.disconnect();
+      }
     }
+
+    if (!isResponseSuccess(responseCode)) {
+      throw getExceptionForHttpResponse(responseCode);
+    }
+
     return parseNewRelease(responseBody);
   }
 
-  AppDistributionReleaseInternal parseNewRelease(String responseBody)
+  private String readResponseBody(HttpsURLConnection connection) throws IOException {
+    boolean isSuccess = isResponseSuccess(connection.getResponseCode());
+    InputStream is = isSuccess ? connection.getInputStream() : connection.getErrorStream();
+    if (!isSuccess && connection.getErrorStream() == null) {
+      // If the server returns an empty response with an error code, getErrorStream returns null
+      return "";
```

Comment:
Because it feels like a good representation of a an empty response body, which is what it means when `getErrorStream()` returns null and the response code is non-2XX. I've updated to try and clarify this a bit.